### ISSUE TITLE
libjpeg turbo migration

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/conda-forge-pinning-feedstock#348]
Issue:

We've been discussing moving to libjpeg-turbo for a few reasons:
1. It is faster than `jpeg`
2. It is necessary for use of tiffffile's C accelerated functions.

Unfortunatley, this isn't an easy migration, and I don't know how to make the PR for it myself.

Help would be appreciated to migrate conda-forge to using it. I think @jakirkham suggested having builds that support both `jpeg` and `libjpeg-turbo`.

xref: https://github.com/conda-forge/staged-recipes/pull/6842

**Comment by isuruf** [comment]
Have a look at how the blas switching is done and you can do the same here as jpeg 8 and libkpeg-turbo are ABI compatible.

**Comment by hmaarrfk** [comment]
man that recipe is pretty complicated and I've been having trouble adding variants to my package all day.... not fun.

This is the recipe I should be looking at correct?
https://github.com/conda-forge/blas-feedstock/blob/master/recipe/meta.yaml

**Comment by isuruf** [comment]
Here are the things to check.

1. Do they have the same set of symbols?
2. Do they have the same SONAMEs?
3. Do they have the same compatibility-version in OSX?

**Comment by hmaarrfk** [comment]
I guess I just couldn't get a selector like ` [include_myfeature == true]` to work. so basic. but so frustrating.

**Comment by sdvillal** [comment]
See also comments in https://github.com/conda-forge/staged-recipes/pull/10328

**Comment by hmaarrfk** [comment]
Also relevant for libjpeg turbo12

https://github.com/libjpeg-turbo/libjpeg-turbo/issues/199

**Comment by hmaarrfk** [comment]
@sdvillal thanks so much for your detailed posts. I know exactly where you are coming from. 

https://github.com/libjpeg-turbo/libjpeg-turbo#what-about-libjpeg-v9

Crawling through your posts, I notice that the v9 API is not supported by Turbo and their explination above makes it seem like they won't be implementing it any time soon.

**Comment by isuruf** [comment]
I'm okay with downgrading to v8 API, and have a install-time switching like BLAS if they are ABI compatible.

**Comment by hmaarrfk** [comment]
I love this comment: https://github.com/loopbio/opencv-feedstock/issues/9#issuecomment-370575418

**Comment by sdvillal** [comment]
Yeah, I lost my soul there, but I did it for a [good reason](http://blog.loopbio.com/video-io-2-jpeg-decoding.html).

Still I will try to give a hand here. @hmaarrfk ping me if we get stuck somewhere.

**Comment by hmaarrfk** [comment]
I would like us to reconsider the plan of having two implementations of jpeg around:

My justifications that other linux distros have simply moved on to turbo:

1. Fedora seems to only package `libjpeg-turbo`: https://packages.fedoraproject.org/search?query=libjpeg
2. Ubuntu seems to only package `libjpeg-turbo`: it provides a virtual package for libjpeg8 pointing to libjpeg-turbo
3. Arch seems to only package `libjpeg-turbo`: https://archlinux.org/packages/?sort=&q=libjpeg&maintainer=&flagged=

as such, I don't see the need support two jpeg. @sdvillal 's experiments (and it seems many other distros) also take libjpeg-turbo to be their reference implementation.

Beyond the fact that blas is really 

My main concern is that the conda solver will not play nicely to version `8` being better than `9`. I've seen this many times with numpy being pinned at a slightly lower than recent version by some subpackage in my environment. I would have to tell my package that I wanted my new version of the package, at the expense of numpy. Sometimes rebuilding my environment seemed like the easier path.

To ensure that things don't break in this migration, I would like to propose the following migration path
1. Add libjpeg v8 branch to  the ABI migration to ensure the bot takes care of it.
2. Update libjpeg's feedstock v8 to be up todate with our April 2022 standards. 
3. Add a `run_constraint` on libjpeg turbo to depend on `libjpeg >999999999` so that they can't both be installed at once.
4. Add a migrator to build out BOTH version 9 and version 8. This will help identify the repos that will need massaging to work with turbo before we start a migration.

I think the above steps would be common to any plan.

# JPEG-turbo only plan
6. Add a migrator to rename `jpeg` to `libjpeg-turbo`
7. Archive jpeg-feedstock and move all builds to broken so that they aren't used in figure recipes.

# JPEG + JPEG-turbo plan

6. Create a metapackage `libjpeg` akin to `blas` which pulls `jpeg` or `libjpeg-turbo` as needed.
7. Create a migrator to rename `jpeg` to `libjpeg`.
8. Maintain the following until we decide otherwise:
     * jpeg 9
     * jpeg 8
     * libjpeg-turbo


ultimatley, it is somewhat annoying to deal with this fork, I wish there was a way we could simply have both installed in the same environment.

**Comment by hmaarrfk** [comment]
closed as duplicate of https://github.com/conda-forge/conda-forge.github.io/issues/673#issuecomment-1426881229
